src/main/java/uk/ac/cam/cl/dtg/segue/api/userAlerts/UserAlertsWebSocket.java

```diff
@@ -43,12 +55,34 @@
     private Session session;
     private static ObjectMapper objectMapper = new ObjectMapper();
 
-    public static ConcurrentHashMap<Long, ConcurrentLinkedQueue<UserAlertsWebSocket>> connectedSockets = new ConcurrentHashMap<>();
-    private static Long websocketsOpened = 0L;
-    private static Long websocketsClosed = 0L;
+    private static final int MAX_CONCURRENT_WEB_SOCKETS_PER_USER = 10;
+
+    // Named unsafeConnectedSockets because, although non-aggregate operations on the concurrent hash map are fine,
+    // operations on the user sets of websockets are unsafe unless used with the matching user lock.
+    private static Map<Long, Set<UserAlertsWebSocket>> unsafeConnectedSockets = Maps.newConcurrentMap();
+    private static final int MAX_NUMBER_OF_CONCURRENT_USER_TAB_OPERATIONS = 200; // ~ max number of concurrent users
+    private static Striped<Lock> userLocks = Striped.lazyWeakLock(MAX_NUMBER_OF_CONCURRENT_USER_TAB_OPERATIONS);
 
     private static final Logger log = LoggerFactory.getLogger(UserAlertsWebSocket.class);
-    private static final int MaxConcurrentWebSocketsPerUser = 10;
+
+    /**
+     * This static method obtains a user lock and sends an alert to each of that user's websockets.
+     * @param userId ID of the user to send the messages, we do not check its validity here.
+     * @param alert the alert to send to the user.
+     */
+    public static void notifyUserOfAlert(final long userId, final IUserAlert alert) {
+        Lock userLock = userLocks.get(userId);
```

Comment:
Are there any reasons by we would want this to be a distributed lock across multiple instances of the API? Just thinking we did the distributed locking stuff for the events manager in postgres.

Reply:
Yes, my first thought on this was the event management postgres lock.

The locks are used when interacting with the map of websocket connections.
As the map of connections is stored in the Java process's memory (i.e. not shared between multiple instances of the API) there is currently no benefit to having the lock distributed.

The striped locks are a little more lightweight than having to get postgres involved.

To get the websocket messaging working with multiple APIs we would need the map of websockets to be shared between the various APIs using redis/memcached or we might be able to load balance connections in a sticky way so that every one of a user's websocket connections are being handled by the same API.

Whichever way we decide to do it, it would need some rethinking and rejigging, and I'm pretty sure we'd reconsider the locking when adapting the map to work with multiple APIs. I will add a comment about using postgres for the distributed lock if we get to having multiple APIs processing websockets correctly at the same time.